### ISSUE TITLE
Making RTM.SendMessage() non-blocking

### ISSUE THREAD

**Issue by Helcaraxan** [slack-go/slack#237]
While looking at using this library (looks great ❤️ ) in a project of mine that requires the use of the websocket API, I was wondering if there was a specific reason for which the [`RMT.SendMessage()`](https://github.com/nlopes/slack/blob/107290b5bbaf3e634833346bb4ff389b1c782bc7/websocket.go#L92) is blocking?

From a code-perspective this is due to the way [`RTM.handleEvents()`](https://github.com/nlopes/slack/blob/107290b5bbaf3e634833346bb4ff389b1c782bc7/websocket_managed_conn.go#L166) and [`RTM.sendOutgoingMessage()`](https://github.com/nlopes/slack/blob/107290b5bbaf3e634833346bb4ff389b1c782bc7/websocket_managed_conn.go#L232) are implemented, _i.e_ in a non-concurrent way.

I personally would find it worrying, especially when creating for example a server / chatbot, to have to deal with network lag & flakiness at my application level to prevent my server from stuttering, etc. I would intuitively think that would be something a library should deal with.

If making this piece of the code more robust / concurrent / scalable is something that is of interest to people, and not in conflict with any overall philosophy I would be more than happy to contribute the necessary work & PRs in this direction. 🙂 

Thought I would rather ask first before starting a larger piece of work and random PRs. 😉

**Comment by dvdplm** [comment]
My two cents on this: I think the Slack library should avoid having opinions about how people structure their apps and write their code and instead concentrate on providing ergonomic primitives for them to do so. In that spirit, it seems to me like there's an advantage in letting the caller do `go RTM.SendMessage(…)` if they wish (and take on the responsibility to ensure no params passed are written to by other threads).
What are the arguments for making the call non-blocking in your opinion? Or rather, what's bad about prefixing `go` in your app?

**Comment by Helcaraxan** [comment]
Even using
```
go RTM.SendMessage()
```
will still be kind-of blocking and only shove the ugliness under the carpet. These calls involve going through a non-buffered channel [here](https://github.com/nlopes/slack/blob/107290b5bbaf3e634833346bb4ff389b1c782bc7/websocket.go#L98). As said this means your go-routine should actually be handling time-outs, and in the case of heavy-use this means you might get 100s of blocking go routines (haven't followed through entirely but it might even be that this could possibly `panic` in edge-cases if the connection is shutdown).

It also means that every user of this library that cares somewhat about dealing with re-sending undelivered messages after a transient connection failure will need to implement their own custom version of what I am proposing here. I definitely can agree that this does not concern all users but I would argue that it is not just a one-in-a-million situation.

Also, if `nlopes/slack` can take care of this with clear and properly defined semantics, it should also not hinder those users that don't care whether undelivered messages are resent or not.

**Comment by zellyn** [comment]
In general, the guidance I've seen in Go is to favor a simple synchronous API and let callers spin off goroutines if they want to, as @dvdplm mentioned. However, your points about handling retries and timeouts in the context of connection failovers are good, and make this case a bit different: it would be really nice not to have to worry about those.

**Comment by Helcaraxan** [comment]
@zellyn I definitely agree about the "favor simple synchronous APIs".

But indeed in this case I consider it to be more a question of cross-project refactoring than one of "keep-the-API-simple". The functionality that I am suggestion has a very narrow specification and clear target. It is not really a topic on which every project using `nlopes/slack` will have its varying and separate opinion:
- You either want to fire-and-forget your messages, in which case you don't care if you have a guaranteed delivery or not.
- Or you want some form of guarantee on your delivery and the ability to check whether something was delivered yet, in which case you'd rather have it out-of-the-box rather than having to code your own ack-or-resend mechanism with associated message queue and logic, etc.

**Comment by Helcaraxan** [comment]
....
Continuing on the conversations started in #238 around what to do with the RTM / WS part of `nlopes/slack`. cc @dvdplm, @james-lawrence, @zellyn.

---

_[By @james-lawrence](https://github.com/nlopes/slack/pull/238#issuecomment-356719252)_
> I believe it falls in line with the utility functions we discussed in another PR, outside of scope for the core, but definitely has utility and can be add in a subpackage.

The suggested changes / features here are unfortunately not really "add-on". They cannot be built on top of the existing low-level WS handling but would require a rewrite of that piece of code. This makes the ability to provide them in a sub-package relatively unlikely.

---

_[By @dvdplm](https://github.com/nlopes/slack/pull/238#issuecomment-356705969)_
>This is slightly off topic here, please bear with me.
I've always felt that the managed WS code is something of an odd duck in this library – it's quite different in scope and purpose of most other parts of the code. How about removing it? I.e. move it to a separate "slack-ws" project? The would be a breaking change of course, but switching WS libs is also sort of a breaking change right? It feels like a separate project and repo could evolve much faster and provide more features, e.g. like outlined in #237
>
> Thoughts?

Acknowledging, but putting temporarily aside the fact that I'm just the new guy in town (_i.e_ I have no long-term context and will defer to the core contributors): it does sound like this is something possible and I agree that it _might_ speed-up development.

The RTM and Event APIs are sufficiently disjoint that I don't think it would be a huge problem even from a user perspective. Even if one would like to combine both in a single project it would still be possible to depend on both the reduced `nlopes/slack` and the theoretical `slack-ws` project.

A potential alternative, maybe in combination with other things planned for the `2.0` version(*), is to keep all the API struct definitions (shared or not between Events and RTM APIs) in the top-level package and split the logic of the Events API and the RTM API in two different sub-packages. This would have the benefit of keeping a single "Slack" library while providing a clearer split in the implementations with a greater (perceived) ability to speed-up development on both sides.

(*) Can't seem to find much context on the `2.0` topic in issues / docs / PRs - any pointers I can have a look at planning / feature / timeline wise?

**Comment by james-lawrence** [comment]
tl;dr; async doesn't actually solve the problems mentioned, will add surprising behaviour and makes error handling more difficult. I'm a strong no going async given current information being discussed.

@Helcaraxan one question:

why? your claims don't really hold up.

> I personally would find it worrying

great, why? have you actually run into any issues that actually require async and can't be solved by a simple automatic retry?

> when creating for example a server / chatbot, to have to deal with network lag & flakiness at my application level to prevent my server from stuttering, etc.

latency/api issues are going to cause your message delivery to have issues no matter what is done by this library.

async does not solve any of the above issues, instead the situation will have been made worse as  async will cause:

- messages to be delivered out of order by the library.
- messages to be dropped on the floor.
- handling error conditions now becomes harder for users to deal with.

all net losses for users of this library.

> If making this piece of the code more robust / concurrent / scalable.

these are claims without actual proof of a problem. and regardless async will not likely solve any one them.

- robust, see comments above on out of order delivery and error handling.
- concurrent, scalable. given slack rate limits their endpoints I doubt these actually come up in practice. 

> The suggested changes / features here are unfortunately not really "add-on". They cannot be built on top of the existing low-level WS handling but would require a rewrite of that piece of code. This makes the ability to provide them in a sub-package relatively unlikely.

You're assuming I think the current WS code is what should belong in the core package.

regarding 2.0:

I'd prefer to do it sooner rather than later as to get alot of the old cruft out of the global namespace of the package. but I'm happy to let the idea sit to see what comes up that we can't currently do without breaking the API.

**Comment by Helcaraxan** [comment]
I take note of the preference for not using any async behaviour and have no problems with that outcome. This discussion and your points and questions were exactly the reason why I raised this issue: to clear those things up before I spent time coding something that people do not agree with. 🙂 

Even if I do not necessarily agree with the points you have raised, and am happy to discuss them further, I think it is more useful now for the library's sake if I pick up less controversial issues that are still open and help to address them if I can.

**Comment by james-lawrence** [comment]
@Helcaraxan thanks for raising the issue. I just don't think we have any justifiable reason for the complexity beyond what ifs and hypotheticals.

if the other maintainers think there is value here, or we see some normal behaviour that is hard to deal with outside of the library then I'll happily review my current stance. =)

**Comment by james-lawrence** [comment]
@Helcaraxan one possible alternative would be to make the RTM's send channel buffer adjustable. would that solve your usecase?
